src/main/java/org/opensearch/knn/index/quantizationService/QuantizationService.java

```diff
@@ -0,0 +1,128 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package org.opensearch.knn.index.quantizationService;
+
+import lombok.AccessLevel;
+import lombok.NoArgsConstructor;
+import org.apache.lucene.index.FieldInfo;
+import org.opensearch.knn.index.VectorDataType;
+import org.opensearch.knn.index.engine.qframe.QuantizationConfig;
+import org.opensearch.knn.index.vectorvalues.KNNVectorValues;
+import org.opensearch.knn.quantization.factory.QuantizerFactory;
+import org.opensearch.knn.quantization.models.quantizationOutput.BinaryQuantizationOutput;
+import org.opensearch.knn.quantization.models.quantizationOutput.QuantizationOutput;
+import org.opensearch.knn.quantization.models.quantizationParams.QuantizationParams;
+import org.opensearch.knn.quantization.models.quantizationParams.ScalarQuantizationParams;
+import org.opensearch.knn.quantization.models.quantizationState.QuantizationState;
+import org.opensearch.knn.quantization.quantizer.Quantizer;
+import java.io.IOException;
+
+import static org.opensearch.knn.common.FieldInfoExtractor.extractQuantizationConfig;
+
+/**
+ * A singleton class responsible for handling the quantization process, including training a quantizer
+ * and applying quantization to vectors. This class is designed to be thread-safe.
+ *
+ * @param <T> The type of the input vectors to be quantized.
+ * @param <R> The type of the quantized output vectors.
+ */
+@NoArgsConstructor(access = AccessLevel.PRIVATE)
+public final class QuantizationService<T, R> {
```

Comment:
The meaning of Service is pretty different in Opensearch. This is more of a manager and not service. If it is service then why it is has generics in it.